solarkraft/test/e2e/verify.test.ts

```diff
@@ -85,31 +93,62 @@ describe('verify', () => {
-        '337365516cbe5f9287bc6b81428fb94be08d8e5d1ba580454d0907378c06c0da',
-        'ad3c4d82300b50972fdbc887c20bb7771c722a1c63bd44c8e8d4ddf4597f4143',
-        'e3dba60ada9fd7cb1d3490cf72b5e53e17b154e237aed2d2dcacfec6b0b1d883',
-        '762575ca214bd9cb85824356b799689bf4920aafcac24f790defac54b1a0cc9d',
-        '98c2468aad6626b59230b4fcc83d921897cbf76dc49f7bea13e46e443035a682',
-        'df673ab5399dcbd5ac434c0ac51c53e9b9367b1d2507ab7e4b8cebfe3e796499',
-        '101111c4a7087070366d035f59cc20b80215231955f9253c4ad85ff94f1dc61b',
-        '9c46c29a985ceeaffe567dc405a757805ff0496fe5740228abcb22b7aefa3e64',
-        '4da2e37effde690b6cf2c11ade5fc2f4e72971b37866f6d75fa2ea5728b6be6e',
-        'b39cf7d57120fa72bde0aec90f709551ff777132e1b5db89e2575ee942d8bb70',
-        '1bd53234a4eb4f9316f63bd77f4a2c191bded377ad3c53caeb7ed285f9d77d64',
-        '406d278860b5531dd1443532f3457c5daa288e8eb0007d2a8e2aa0127e87949e',
-    ]
-    txs.forEach((tx) => {
-        it(`verifies the setter contract (tx ${tx})`, function (done) {
-            this.timeout(150_000)
-            spawn(
-                `solarkraft verify --home test/e2e/tla/ --txHash ${tx} --monitor test/e2e/tla/setter_mon.tla`
-            )
-                .wait('[OK]')
-                .run(done)
-        })
+    it('fetches and verifies', function (done) {
+        // fetch the transactions like in fetch.test.ts
+        const solarkraftHome = './test/e2e/tla'
+        this.timeout(120000)
+
+        // this callback is called after the fetch command has finished
+        const fetchCallback = (err) => {
+            assert(!err, `fetch failed: ${err}`)
+            // verify all of the transactions
+            let ncalls = 0
+            let nchecked = 0
+            for (const e of yieldListEntriesForContract(
+                SETTER_CONTRACT_HASH,
+                join(solarkraftHome, '.stor', SETTER_CONTRACT_HASH)
+            )) {
+                spawn(
+                    'solarkraft',
+                    [
+                        'verify',
+                        `--home=${solarkraftHome}`,
+                        `--txHash=${e.txHash}`,
+                        '--monitor=./test/e2e/tla/setter_mon.tla',
+                    ],
+                    { verbose: true }
+                )
+                    .wait('[OK]')
+                    .run((err) => {
+                        assert(!err, `verify failed: ${err}`)
+                        nchecked++
+                        console.log(`checked ${e.txHash} of ${ncalls}`)
+                        if (nchecked >= ncalls) {
+                            // technically, we may have a race condition
+                            // between spawning and finishing verification
+                            assert(ncalls > 0, 'no transactions checked')
+                            // we have to tell mocha that we are done
+                            done()
+                        }
+                    })
+                ncalls++
```

Comment:
This looks super fishy. You're incrementing `ncalls` outside the callback.
Won't you call `done()` early if the callback executes before the outer loop makes another call?
It looks like you're assuming that the outer loop executes faster than the callback, but it may be completely arbitrary.

I think the proper way to do this is to return a promise and wait for `Promise.all()`.

Reply:
yes, I was fighting with this a lot. You are right. Ideally, we would like to wait for all the promises to end. However, `nexpect` is not using promises. Instead, it uses some kind of `node` tasks. I have no idea how to wait there. So it's the best I could come up with. It may terminate earlier.

Reply:
I have found a way to do it properly, similar to how you did it by dynamically introducing tests in a loop. The same trick works with a `before` hook!